File: src/main/java/ru/pearx/pmdumper/dumpers/DumperCapabilities.java
package ru.pearx.pmdumper.dumpers;

import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityManager;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.apache.commons.lang3.reflect.FieldUtils;
import ru.pearx.pmdumper.PMDumper;
import ru.pearx.pmdumper.dumpers.base.IDumper;
import ru.pearx.pmdumper.dumpers.base.PMDData;

import java.util.*;

/**
 * Created by mrAppleXZ on 12.06.17 11:23.
 */
@SideOnly(Side.CLIENT)
public class DumperCapabilities implements IDumper
{

    @Override
    public String getName()
    {
        return "capabilities";
    }

    @Override
    public PMDData getData()
    {
        List<List<String>> lst = new ArrayList<>();
        lst.add(Arrays.asList("Interface", "Default Instance Class", "Storage Class"));
        try
        {
            IdentityHashMap<String, Capability<?>> map = (IdentityHashMap<String, Capability<?>>) FieldUtils.readField(CapabilityManager.INSTANCE, "providers", true);
            for (Map.Entry<String, Capability<?>> entr : map.entrySet())
            {
                     lst.add(Arrays.asList(entr.getKey(), entr.getValue().getDefaultInstance().getClass().getName(), entr.getValue().getStorage().getClass().getName()));
            }
        } catch (IllegalAccessException e)
        {
            PMDumper.INSTANCE.log.error(e);
        }
        return new PMDData(lst, null);
    }
}

File: src/main/kotlin/ru/pearx/pmdumper/dumper/standard/DumperCapabilities.kt
@file:JvmMultifileClass
@file:JvmName("StandardDumpers")

package ru.pearx.pmdumper.dumper.standard

import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityManager
import ru.pearx.pmdumper.ID
import ru.pearx.pmdumper.dumper.dumper
import ru.pearx.pmdumper.utils.add
import ru.pearx.pmdumper.utils.readField
import ru.pearx.pmdumper.utils.tryDump
import java.util.*

val DumperCapabilities = dumper {
    registryName = ResourceLocation(ID, "capabilities")
    header = listOf("Interface", "Default Instance Class", "Storage Class")
    iterator {
        for ((key, value) in CapabilityManager.INSTANCE.readField<IdentityHashMap<String, Capability<*>>>("providers")) {
            tryDump(ArrayList(header.size)) {
                add(key)
                add {
                    val defaultInstance = value.defaultInstance
                    if (defaultInstance == null) "" else defaultInstance::class.java.name
                }
                add(value.storage::class.java.name)
            }
        }
    }
}